plugins/synaptics-rmi/fu-synaptics-rmi-device.c

```diff
@@ -0,0 +1,1098 @@
+			fu_byte_array_append_uint8 (req, 0x0);
+		if (g_getenv ("FWUPD_SYNAPTICS_RMI_VERBOSE") != NULL) {
+			fu_common_dump_full (G_LOG_DOMAIN, "ReportWrite",
+					     req->data, req->len,
+					     80, FU_DUMP_FLAGS_NONE);
+		}
+		if (!fu_io_channel_write_byte_array (priv->io_channel, req, RMI_DEVICE_DEFAULT_TIMEOUT,
+						     FU_IO_CHANNEL_FLAG_SINGLE_SHOT |
+						     FU_IO_CHANNEL_FLAG_USE_BLOCKING_IO, error))
+			return NULL;
+
+		/* response */
+		res = fu_io_channel_read_byte_array (priv->io_channel, req_sz, RMI_DEVICE_DEFAULT_TIMEOUT,
+						     FU_IO_CHANNEL_FLAG_NONE, error);
+		if (res == NULL)
+			return NULL;
+		if (g_getenv ("FWUPD_SYNAPTICS_RMI_VERBOSE") != NULL) {
+			fu_common_dump_full (G_LOG_DOMAIN, "ReportRead",
+					     res->data, res->len,
+					     80, FU_DUMP_FLAGS_NONE);
+		}
+		if (res->len < HID_RMI4_READ_INPUT_DATA) {
+			g_set_error (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INTERNAL,
+				     "response too small: 0x%02x",
+				     res->len);
+			return NULL;
+		}
+		input_count_sz = res->data[HID_RMI4_READ_INPUT_COUNT];
+		if (input_count_sz == 0) {
+			g_set_error_literal (error,
+					     FWUPD_ERROR,
+					     FWUPD_ERROR_INTERNAL,
+					     "input count zero");
+			return NULL;
+		}
+		if (input_count_sz < (guint) req_sz) {
+			g_set_error (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INTERNAL,
+				     "input count smaller 0x%02x than request 0x%02x",
+				     input_count_sz, (guint) req_sz);
+			return NULL;
+		}
+		if (input_count_sz + (guint) HID_RMI4_READ_INPUT_DATA > res->len) {
+			g_set_error (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INTERNAL,
+				     "underflow 0x%02x from expected 0x%02x",
+				     res->len, (guint) input_count_sz + HID_RMI4_READ_INPUT_DATA);
+			return NULL;
+		}
+		g_byte_array_append (buf,
+				     res->data + HID_RMI4_READ_INPUT_DATA,
+				     input_count_sz);
+	}
+	if (g_getenv ("FWUPD_SYNAPTICS_RMI_VERBOSE") != NULL) {
+		fu_common_dump_full (G_LOG_DOMAIN, "DeviceRead", buf->data, buf->len,
+				     80, FU_DUMP_FLAGS_NONE);
+	}
+
+	return g_steal_pointer (&buf);
+}
+
+gboolean
+fu_synaptics_rmi_device_write (FuSynapticsRmiDevice *self, guint16 addr, GByteArray *req, GError **error)
+{
+	FuSynapticsRmiDevicePrivate *priv = GET_PRIVATE (self);
+	guint8 addr_le[2] = { 0x0 };
+	guint8 len = 0x0;
+	g_autoptr(GByteArray) buf = g_byte_array_new ();
+
+	/* check size */
+	if (req != NULL) {
+		if (req->len > 0xff) {
+			g_set_error_literal (error,
+					     FWUPD_ERROR,
+					     FWUPD_ERROR_INTERNAL,
+					     "data to write was too long");
+			return FALSE;
+		}
+		len = req->len;
+	}
+
+	/* report */
+	fu_byte_array_append_uint8 (buf, RMI_WRITE_REPORT_ID);
+
+	/* length */
+	g_byte_array_append (buf, &len, sizeof(len));
+
+	/* address */
+	fu_common_write_uint16 (addr_le, addr, G_LITTLE_ENDIAN);
+	g_byte_array_append (buf, addr_le, sizeof(addr_le));
+
+	/* optional data */
+	g_byte_array_append (buf, req->data, req->len);
```

Comment:
Isn't this possibly dereferencing a null pointer with it being optional data?

Reply:
Good catch.